### DISCUSSION TITLE
Is it possible to use the v4 model for an Azure Function based bot application?

### DISCUSSION THREAD

**Discussion by krli-fadv** [OfficeDev/microsoft-365-agents-toolkit#10664] [Q&A]
Hello. I created an Azure Function based bot application by using the Teams Toolkit extension and everything worked fine. However, I wanted to use Azure Functions v4 with it, so I upgraded the @azure/functions package to 4.1.0. I migrated each function I had to use the v4 model, but I ran into some issues with the messageHandler. 

I wasn't sure if it was entirely possible to use the v4 model because the examples seem to be using the v3 model and there seems to be significant changes with the v4 model. Is it possible to use the v4 model for an Azure Function based bot application? If so, are there any recommendations on how to convert it to using the v4 model? Thank you.

**Comment by therealjohn** [comment]
I'm not sure if there are any blocking issues with using v4, but the sdk, templates, and samples are use v3 model. 

What specific issues did you hit?

**Reply by krli-fadv** [reply]
Thanks for your response. In the Azure Function based bot app I created, I have a command handler enabled and an action handler enabled. After I tried converting the messageHandler, I ran into a couple of issues. The first issue that I ran into with that file was with the `const res = new ResponseWrapper(context.res)` line. With the v4 model, res does not exist anymore on the context, so I just ended up temporarily passing new HttpResponse() as the response for the ResponseWrapper.

After that, I had some issues with passing in the request to notificationApp.requestHandler, but I was able to get around that issue by creating an object with the expected structure for the request and passing it in as the first argument. With those changes, after I entered a command, I was able to see the expected adaptive card, but if I interacted with a button on the card, it wouldn't show the updated adaptive card. However, I do see the message saying "Your response was sent to the app". I'm assuming the issue has to do with what I'm passing into the ResponseWrapper, but I wasn't sure what I could provide to it in this case.
